plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpEmulationTab.java

```diff
@@ -0,0 +1,237 @@
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
+import org.eclipse.debug.ui.AbstractLaunchConfigurationTab;
+import org.eclipse.jdt.internal.ui.viewsupport.ImageDisposer;
+import org.eclipse.jface.layout.GridDataFactory;
+import org.eclipse.jface.layout.GridLayoutFactory;
+import org.eclipse.jface.viewers.ISelectionChangedListener;
+import org.eclipse.jface.viewers.SelectionChangedEvent;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.ModifyEvent;
+import org.eclipse.swt.events.ModifyListener;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.ui.PlatformUI;
+
+public class GcpEmulationTab extends AbstractLaunchConfigurationTab {
+
+  private static final Logger logger = Logger.getLogger(GcpEmulationTab.class.getName());
+
+  private static final String ATTRIBUTE_ACCOUNT_EMAIL =
+      "com.google.cloud.tools.eclipse.gcpEmulation.accountEmail"; //$NON-NLS-1$
+
+  private static final String ATTRIBUTE_GCP_PROJECT =
+      "com.google.cloud.tools.eclipse.gcpEmulation.gcpProject"; //$NON-NLS-1$
+
+  private static final String ATTRIBUTE_SERVICE_KEY =
+      "com.google.cloud.tools.eclipse.gcpEmulation.serviceKey"; //$NON-NLS-1$
+
+  private final IGoogleLoginService loginService;
+  private final ProjectRepository projectRepository;
+
+  private AccountSelector accountSelector;
+  private ProjectSelector projectSelector;
+  private Text serviceKeyInput;
+
+  private Image gcpIcon;
+
+  // We set up intermediary models between a run configuration and UI components for certain values,
+  // because, e.g., the account selector cannot load an email if it is not logged in. In such a
+  // case, although nothing is selected in the account selector, we should not clear the email saved
+  // in the run configuration.
+  private String accountEmailModel;
+  private String gcpProjectIdModel;
+  // To prevent updating above models when programmatically setting up UI components.
+  private boolean initializingUiValues;
+
+  public GcpEmulationTab() {
+    this(PlatformUI.getWorkbench().getService(IGoogleLoginService.class),
+        new ProjectRepository(PlatformUI.getWorkbench().getService(IGoogleApiFactory.class)));
+  }
+
+  @VisibleForTesting
+  GcpEmulationTab(IGoogleLoginService loginService, ProjectRepository projectRepository) {
+    this.loginService = loginService;
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void createControl(Composite parent) {
+    Composite composite = new Composite(parent, SWT.NONE);
```

Comment:
I wonder if this, and dataflow, should instead have a read-only representation with an _Edit…_ button that pops this up as a dialog.  That would avoid the dynamic background updates etc.

Reply:
That's an attractive idea. I actually thought about having a read-only text field for the project (the account is just for the project selector), which can display the currently saved project even if the user is not logged in. Then we have a "Select Project..." button next to it, which pops up the project selector dialog. I really like this idea considering the benefits from the implementation perspective. The UI is fast because there won't be a long-running project fetch, no async issue in the run config, and a very clean model in the case of users not logged in. @elharo what do you think of this approach?

Reply:
@elharo likes the idea too, so we can seriously consider going that route. We agreed that we first move forward with this PR (hoping to be able to release this soon) and implement the idea later. I don't think it will be more difficult to do it later than doing it now; it will be changes only in the UI.